### PR TITLE
Generate diagnostics docs

### LINKED ISSUE #6215: Document rust-analyzer diagnostics
We should document all our diagnostics in the manual (probably don't need a lot of text, just the diagnostic code and maybe an example message). We could probably generate this similarly to the assist documentation.

### PR DESCRIPTION
**Pull request by popzxc**
Resolves #6215

### CODE REVIEW

**Comment by kjeremy** [comment]
For fun we could also implement [CodeDescription](https://microsoft.github.io/language-server-protocol/specifications/specification-3-16/#diagnostic) if supported to refer back to these generated docs

File: crates/hir_def/src/diagnostics.rs [hunk 1]
```diff
@@ -50,6 +56,9 @@ impl Diagnostic for UnresolvedExternCrate {
     }
 }
 
+// Diagnostic: unresolved-import
+//
+// This diagnostic is triggered if rust-analyzer is unable to discover imported module.
```
Comment (jonas-schievink) [inline crates/hir_def/src/diagnostics.rs:61]:
```suggestion
// This diagnostic is triggered if rust-analyzer is unable to discover an imported item.
```

File: xtask/src/codegen/gen_diagnostic_docs.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    fn collect() -> Result<Vec<Diagnostic>> {
+        let mut res = Vec::new();
+        for path in rust_files(&project_root()) {
+            collect_file(&mut res, path)?;
+        }
+        res.sort_by(|lhs, rhs| lhs.id.cmp(&rhs.id));
+        return Ok(res);
+
+        fn collect_file(acc: &mut Vec<Diagnostic>, path: PathBuf) -> Result<()> {
+            let text = fs::read_to_string(&path)?;
```
Comment (matklad) [inline xtask/src/codegen/gen_diagnostic_docs.rs:36]:
```suggestion
            let text = xshell::read_file(path)?;
```

that'll get a better error message

File: xtask/src/codegen/gen_diagnostic_docs.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+use std::{fmt, fs, path::PathBuf};
+
+use crate::{
+    codegen::{self, extract_comment_blocks_with_empty_lines, Location, Mode, PREAMBLE},
+    project_root, rust_files, Result,
+};
+
+pub fn generate_diagnostic_docs(mode: Mode) -> Result<()> {
+    let features = Diagnostic::collect()?;
```
Comment (matklad) [inline xtask/src/codegen/gen_diagnostic_docs.rs:11]: Copy-paste detected :-)

This is fine for such code, though pulling common "traverse the files and colllect magic comments" might be pulled into some helper function somewhere.

File: xtask/src/codegen.rs [hunk 4]
```diff
@@ -47,6 +49,7 @@ impl CodegenCmd {
         generate_assists_tests(Mode::Overwrite)?;
         generate_assists_docs(Mode::Overwrite)?;
         generate_feature_docs(Mode::Overwrite)?;
+        generate_diagnostic_docs(Mode::Overwrite)?;
```
Comment (matklad) [inline xtask/src/codegen.rs:52]: Let's add this to `smoke_test_docs_generation` as well

Comment (matklad) [reply xtask/src/codegen.rs:52]: Ah, we also need logic in `release.rs` to copy the file over.

File: xtask/src/codegen/gen_diagnostic_docs.rs (hunk 3 above)
Comment (popzxc) [reply xtask/src/codegen/gen_diagnostic_docs.rs:11]: 😊

Comment (popzxc) [reply xtask/src/codegen/gen_diagnostic_docs.rs:11]: Regarding the pulling common code: I think it'll be better to not mix refactoring and new functionality in one PR, if that's possible.

**Comment by matklad** [comment]
bosr r+

Thanks!

Regarding refactorings, for smaller ones like this, I prefer them to be in a separate commit at the start PR, but this is a weak preference

**Pull request merged**